### PR TITLE
[neohub] added automatic discovery of NeoHubs

### PR DESCRIPTION
**Pull request by andrewfg**
This PR adds an mDNS auto discovery service participant to discover NeoHubs

### CODE REVIEW

File: bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Discovers NeoHubs by means of mDNS-SD
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true)
```
Comment (cpmeister) [inline bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java:38]: Components shouldn't be declared immediate anymore.

```suggestion
@Component
```

File: bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Override
+    public String getServiceType() {
+        return "_hap._tcp.local.";
+    }
+
+    @Override
+    public @Nullable DiscoveryResult createResult(ServiceInfo serviceInfo) {
+        @Nullable
```
Comment (cpmeister) [inline bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java:73]: You can omit `@Nullable` for local variables.
```suggestion
```

File: bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    .build();
+            logger.debug("Discovered a NeoHub on host '{}'", ipStr);
+            return hub;
+        }
+        return null;
+    }
+
+    @Override
+    public @Nullable ThingUID getThingUID(ServiceInfo serviceInfo) {
+        @Nullable
```
Comment (cpmeister) [inline bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java:89]:
```suggestion
```

File: bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    public @Nullable DiscoveryResult createResult(ServiceInfo serviceInfo) {
+        @Nullable
+        String ipStr = getIpAddressIfValidNeoHub(serviceInfo);
+        if (ipStr != null) {
+            ThingUID thingUID = new ThingUID(NeoHubBindingConstants.THING_TYPE_NEOHUB, ipStr.replace('.', '_'));
+            DiscoveryResult hub = DiscoveryResultBuilder.create(thingUID)
+                    .withProperty(NeoHubConfiguration.HOST_NAME, ipStr)
+                    .withRepresentationProperty(NeoHubConfiguration.HOST_NAME).withLabel("NeoHub (" + ipStr + ")")
+                    .build();
+            logger.debug("Discovered a NeoHub on host '{}'", ipStr);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubDiscoveryParticipant.java:81]: The discovery service is going to log this anyway, so I don't think this log is necessary.

**Review by cpmeister — CHANGES_REQUESTED** [review]

**Comment by andrewfg** [comment]
Many thanks for the feedback @cpmeister. I made the following changes..

- removed the underlying need for those null annotations
- eliminated the log message (and thus the logger instance)
- made the search filter potentially a bit more future proof

**Review by cpmeister — APPROVED** [review]

**Pull request merged**